### PR TITLE
fix: correctly obtain relative path required for the venv created by `--bootstrap_impl=script`

### PR DESCRIPTION
**Pull request by chowder**
Computing the relative path from the venv interpreter to the underlying interpreter was
incorrectly using the actual interpreter's directory depth, not the venv interpreter's
directory depth, when computing the distance from the venv interpreter to the runfiles root.
The net effect is the correct relative path would only be computed for binaries with
the same directory depth as the actual interpreter (e.g. 2).

This went undetected in CI because the tests for this logic just happen to have the
same directory depth as the actual interpreter used.

To fix, compute the relative path to the runfiles root using the venv interpreter
directory. Also added a test in a more nested directory to test this case.

Along the way:
* Change relative path computation to compute a minimum relative path.
* Fix the internals to pass a runfiles-root relative path, not main-repo relative path,
  for the actual interpreter, as intended.

Fixes https://github.com/bazelbuild/rules_python/issues/2169

### CODE REVIEW

**Comment by ewianda** [comment]
The following rules were tested and validated:

- [x] **py_binary**  
- [x] **py_test**  
- [x] **py_console_script**  
- [x] **rules_oci Python image**

---

**Review by rickeylev — COMMENTED** [review]
Ah, hah. I'm guessing the reason this worked in CI is the path depth of the test programs was the same as the path depth of the underlying interpreter? :joy: 

Thanks for the PR!

The relative path logic looks overly complicated, though? All it has to do is calculate how many up-escapes are needed to get from the `{prefix}.venv/bin/python3` location to the runfiles root, then append the `../`-stripped value of the actual interpreter to that.

I think the `../` prefixed part of e.g. `runtime.interpreter.short_path` (or any `.short_path`) is there in lieu of `external/`, and represents "go up one dir from the main repo" (because its referring to some external repo).

**Comment by chowder** [comment]
Yeah it's kinda complicated (don't like it) - I wasn't sure what were all the edge cases here, so writing it generically was the only way I knew how to do it.

**Comment by rickeylev** [comment]
> would something like this suffice

Yes, it should. I pushed a change that basically did that -- compute the relative path between two runfiles-root-relative paths instead of main-repo-relative paths. I kept the relative_path function since I like that it will compute a minimum relative-path.

I also pushed a test to reproduce the problem. It creates a test in a much more deeply nested directory than where the actual interpreter is. As well as some other cleanup.

**Comment by rickeylev** [comment]
> I wasn't sure what were all the edge cases here

I think you got them all, so good job :slightly_smiling_face: . That `short_path` is a main-repo relative path, and gets `../` prepended for external things is the main edge case. The code was a bit confusing because it _said_ it was passing runfiles-root-relative paths around, but it actually wasn't (`interpreter_actual_path` was the plain `short_path` value, which then got fixed up later), and the some later code was papering over that. I went and fixed that.

File: tests/bootstrap_impls/venv_relative_path_tests.bzl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            to = "../c/d",
+        ),
+    ).equals("../../c/d")
+
+    env.expect.that_str(
+        relative_path(
+            from_ = "../a/b",
+            to = "../../c/d",
+        ),
+    ).equals("../../../c/d")
```
Comment (aignas) [inline tests/bootstrap_impls/venv_relative_path_tests.bzl:44]: Is this correct?
Lets say current dir is `/foo/bar/`
Then `../a/b` maps to `/foo/a/b`
Then `../../c/d` maps to `/foo/c/d`
Then the relative path of `/foo/c/d` with respect to `/foo/a/b` is `../../c/d`

**Comment by rickeylev** [comment]
The CI failure is due to https://github.com/bazel-contrib/bazel_features/issues/82, so I'll force-merge past it.

**Comment by rickeylev** [comment]
Ah, actually, I don't think I can force-merge because I'm not an admin. I'll ask about getting admin back. Otherwise, we can just switch the "upcoming bazel" CI to 8rc2 instead of 8rc3

File: tests/bootstrap_impls/venv_relative_path_tests.bzl (hunk 1 above)
Comment (chowder) [reply tests/bootstrap_impls/venv_relative_path_tests.bzl]: I don't think this was incorrect.

If current dir was `/foo/bar`, then `../../c/d` maps to `/c/d, not `/foo/c/d`

So from `/foo/a/b` to `/c/d` is indeed `../../../c/d`

**Comment by chowder** [comment]
@rickeylev Thanks for the changes. :)

**Comment by ewianda** [comment]
Apologies. Just noticed that `rules_pkg` is failing 

```
  File "/home/ewianda/.cache/bazel/_bazel_ewianda/00aeeab5df0e41ab9e87aa07d33c6a73/sandbox/linux-sandbox/43/execroot/_main/bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar.runfiles/_main/../rules_pkg~/pkg/private/tar/build_tar.py", line 486, in <module>
    main()
  File "/home/ewianda/.cache/bazel/_bazel_ewianda/00aeeab5df0e41ab9e87aa07d33c6a73/sandbox/linux-sandbox/43/execroot/_main/bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar.runfiles/_main/../rules_pkg~/pkg/private/tar/build_tar.py", line 477, in main
    output.add_manifest_entry(entry, file_attributes)
  File "/home/ewianda/.cache/bazel/_bazel_ewianda/00aeeab5df0e41ab9e87aa07d33c6a73/sandbox/linux-sandbox/43/execroot/_main/bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar.runfiles/_main/../rules_pkg~/pkg/private/tar/build_tar.py", line 342, in add_manifest_entry
    self.add_file(entry.src, entry.dest, **attrs)
  File "/home/ewianda/.cache/bazel/_bazel_ewianda/00aeeab5df0e41ab9e87aa07d33c6a73/sandbox/linux-sandbox/43/execroot/_main/bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar.runfiles/_main/../rules_pkg~/pkg/private/tar/build_tar.py", line 111, in add_file
    self.tarfile.add_file(
  File "/home/ewianda/.cache/bazel/_bazel_ewianda/00aeeab5df0e41ab9e87aa07d33c6a73/sandbox/linux-sandbox/43/execroot/_main/bazel-out/k8-opt-exec-ST-d57f47055a04/bin/external/rules_pkg~/pkg/private/tar/build_tar.runfiles/rules_pkg~/pkg/private/tar/tar_writer.py", line 251, in add_file
    with open(file_content, 'rb') as f:
         ^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: 'bazel-out/k8-fastbuild/bin/external/rules_pkg~/pkg/private/tar/_build_tar.venv/bin/python3'

```
